pythnet/pythnet_sdk/src/wire.rs

```diff
@@ -0,0 +1,306 @@
+            t_str:    &'a str,
+
+            // Test `Vec` is serialized to a variable length array.
+            t_vec:              Vec<u8>,
+            t_vec_empty:        Vec<u8>,
+            t_vec_nested:       Vec<Vec<u8>>,
+            t_vec_nested_empty: Vec<Vec<u8>>,
+            t_slice:            &'a [u8],
+            t_slice_empty:      &'a [u8],
+
+            // Test tuples serialize as expected.
+            t_tuple:        (u8, u16, u32, u64, String, Vec<u8>, &'a [u8]),
+            t_tuple_nested: ((u8, u16), (u32, u64)),
+
+            // Test enum serializations.
+            t_enum_unit:    GoldenEnum,
+            t_enum_newtype: GoldenEnum,
+            t_enum_tuple:   GoldenEnum,
+            t_enum_struct:  GoldenEnum,
+
+            // Test nested structs, which includes our PrefixedVec implementations work as we expect.
+            t_struct:   GoldenNested<u8>,
+            t_prefixed: PrefixedVec<u16, u8>,
+        }
+
+        #[derive(serde::Serialize, serde::Deserialize, Debug, PartialEq)]
+        struct GoldenNested<T> {
+            nested_u8:    T,
+            nested_tuple: (u8, u8),
+        }
+
+        #[derive(serde::Serialize, serde::Deserialize, Debug, PartialEq)]
+        enum GoldenEnum {
+            Unit,
+            Newtype(u8),
+            Tuple(u8, u16),
+            Struct { a: u8, b: u16 },
+        }
+
+        // Serialize the golden test value.
+        let golden_struct = GoldenStruct {
+            unit:               (),
+            t_bool:             true,
+            t_u8:               1,
+            t_u16:              2,
+            t_u32:              3,
+            t_u64:              4,
+            t_string:           "9".to_string(),
+            t_str:              "10",
+            t_vec:              vec![11, 12, 13],
+            t_vec_empty:        vec![],
+            t_vec_nested:       vec![vec![14, 15, 16], vec![17, 18, 19]],
+            t_vec_nested_empty: vec![vec![], vec![]],
+            t_slice:            &[20, 21, 22],
+            t_slice_empty:      &[],
+            t_tuple:            (
+                29,
+                30,
+                31,
+                32,
+                "10".to_string(),
+                vec![35, 36, 37],
+                &[38, 39, 40],
+            ),
+            t_tuple_nested:     ((41, 42), (43, 44)),
+            t_enum_unit:        GoldenEnum::Unit,
+            t_enum_newtype:     GoldenEnum::Newtype(45),
+            t_enum_tuple:       GoldenEnum::Tuple(46, 47),
+            t_enum_struct:      GoldenEnum::Struct { a: 48, b: 49 },
+            t_struct:           GoldenNested {
+                nested_u8:    50,
+                nested_tuple: (51, 52),
+            },
+            t_prefixed:         vec![0u8; 512].into(),
+        };
+
+        golden_struct.serialize(&mut serializer).unwrap();
+
+        // The serialized output should be as expected.
+        assert_eq!(
```

Comment:
This isn't the prettiest way to test, and ideally we'd test variations of each value too (I.E checking true and false for bool and empty non-empty strings) but I will come back to do this after hermes is complete because it's a lot more tests to generate.